### ISSUE TITLE
Units of Particle Binning

### ISSUE THREAD

**Issue by arsampath** [SmileiPIC/Smilei#138]
Dear developers,

I use the following syntax to obtain an energy spectrum of a 2D simulation (y axis: Number of particles in a particular energy range, x axis: energy range  ) in SMILEI.
It is not clear for me how to convert the y axis (the 'data' deposited by ParticleBinning diagnostics) into macro/real number of particles. 

It is given that, In the ParticleBinning post-processing, the results are divided by the hypervolume corresponding to the diagnostic’s definition. So the weight which is in units of NrLr^D becomes NrLr^D/(Lbin^D) where Lbin would be the hypervolume corresponding to the diagnostic. Have I understood your explanation correctly? 

DiagParticleBinning(
    deposited_quantity = "weight",
     every = 1,
     species = ["beamelectron"],
     axes = [["ekin", 190, 200., ncellsx] ] )

data = S.ParticleBinning(0).getData(0)[0] . # Should be now NrLr^D/Lbin^D ?
ekin = S.ParticleBinning(0).get(0)['ekin']

I tried two things: data(Lbin^D/Lr^D)  and data(Lbin^D) to get the normalized number density, but it did not work for me.
How do I convert this extracted data to the number of real/macroparticles?
Thank you for your time.  
I use the latest github version of smilei.

**Comment by mccoys** [comment]
Your variable `data` is indeed in units of `NrLr^D/Lbin^D` which is a number density of real particles (in code units, i.e. ε0 me ωr^2/e^2). You probably want to convert `data` in another units, like SI.

Now, in order to recover a number of real particles, you need to choose a volume `V` inside which the density is constant. Then `data * V` is the number of real particles inside `V`.

If your density is not constant, you should make an integral.

**Comment by mccoys** [comment]
Actually I was not very precise.

`data` is in units of `Nr Lr^D/Vbin`

`Vbin` is a length in 1D, a surface in 2D, and a volume in 3D. This means you always obtain `data` in units of density. Note that `Vbin` is NOT `Lbin^D`.

**Comment by arsampath** [comment]
Thank you for the response. 
In the following input file, I have a cold beam with a constant density profile. I have written the Particlebinning diagnostic I mentioned before. 

[test.txt](https://github.com/SmileiPIC/Smilei/files/3364995/test.txt)

The expected ekin is ~196 mc2 for all the particles. This is seen in the spectrum (S.ParticleBinning(0).plot(0)).

The expected normalized number density is 1.108e-7. (All the particles are supposed to have the same energy since it is a cold beam).

Is the data for this particle binning also in units of density? I see from the graph that it is N_r/K_r which is density by energy. 
When I extract the data, as expected there is only data in one bin, and the value is ~ 2.77e-07. What density does this data correspond to? 
Here the length of each bin, de ~ 0.0098. I tried to process this data by  data * de * de. 
Since you mentioned, the Vbin is not Lbin^D, what does the volume correspond to in this simulation? 

In short, is it possible to know the number of particles with a particular energy using this diagnostic? 

Thank you for your time.

**Comment by mccoys** [comment]
In the case of ekin, you should not multiply by `de*de` but simply `de` because it is not a length.

Then you obtain a number density of real particles.

To obtain the number of real particles, then, the same remark as before applies. You must choose the volume in which you consider your plasma relevant. Remember that, in a PIC code, you obtain densities, not numbers of particles. Only in 3D or is possible to recover an exact number of particles.

**Comment by arsampath** [comment]
Dear developers,

Thank you for your response. 
I would like to know the normalization parameter of the quantity that we obtain upon the integration of the energy distribution function generated using the particle binning diagnostics.

I did the following,
I have an electron beam of finite size and constant density 1 in two dimensions propagating with a relativistic factor of 196. 
Using the particle binning diagnostics, I extracted the weight of the electron beam species as a function of ekin. The ekin range was from 190 to 200 distributed in 10 bins.
Now, I integrated the particle binning data at T = 0 over the entire energy range. 
I obtain a value of N = 0.04882815.
I was expecting this value to be 1 multiplied by the beam volume, which would be 100 in my case as my beam extends for a length of 10 plasma skin depths along both the x and y directions. To check further, I halved as well as doubled the volume of my beam. The value of N proportionally halved and doubled in both the cases as expected. However, the numbers were not clear. Furthermore, I performed simulations with varying particle per cell, density and gamma of the beam. Each time the results were consistent but the number N seemed to be normalised with an arbitrary value of 2048. 

If this N= data*de corresponds to a number density, it should return the beam density (which is equal to 1 Nr)  in all cases, but it does not.

I am attaching my input file with this message. 
[test_pb.txt](https://github.com/SmileiPIC/Smilei/files/3368261/test_pb.txt)
It would be very helpful to me if you kindly let me know the exact normalization used in these computations. In addition, the variation of the results from my expected values by a constant factor of 1/2048 is something that I am not able to interpret.
Kindly help me in this regard.

**Comment by mccoys** [comment]
Your resulting density is an average density over the whole box. Indeed, your diagnostic does not have any spatial resolution, so that it cannot distinguish between the beam and empty regions.

Your box size is `64*32` and your beam size is `10*10`. Consequently, the averaged density is modified by a factor `10*10 / (64*32) = 0.048828125`

**Comment by arsampath** [comment]
Dear developers,

We performed several checks both in 2D and 3D to physically understand the quantity obtained upon the integration of the data set obtained (the weight in our case) over the running variable (ekin). We found that the value obtained is the beam density times the beam volume divided by the system volume.
In our 2D simulations, the system volume was 2048 and hence we found that the values obtained were always divided by a factor of 2048. When we changed the system size, the results obtained were consistent with our interpretation.
This issue seems to be clear now. Hence, I would like to close it.

Thanks
